sui_programmability/examples/games_with_chance/sources/satoshi_flip.move

```diff
@@ -0,0 +1,300 @@
+        option::fill(&mut game.player, tx_context::sender(ctx));
+    }
+
+    /// Called by the house either after `fun bet` has been called by some player to end the game,
+    /// either before bet was called to "cancel" the game.
+    ///
+    /// If the house has "forgotten" the secret, then it can use a random secret and it will
+    /// cancel the game, as long as `fun bet` has not been called.
+    public entry fun end_game(game: &mut Game, secret: vector<u8>, ctx: &mut TxContext) {
+        // only house should be able to end the game
+        assert!(game.house == tx_context::sender(ctx), ENotAllowedToEndGame);
+
+        // house wants to cancel the current game (maybe forgot the secret)
+        if (option::is_none(& game.bet_data)) {
+            let msg = b"Game Canceled by House";
+            let outcome = Outcome {
+                secret: msg,
+                guess: 2,
+                player_won: false
+            };
+            option::fill(&mut game.outcome, outcome);
+
+            // no bet placed, return the balance to house
+            let HouseData {house_balance, min_bet: _, max_bet: _} = option::extract(&mut game.house_data);
+            let house_coins = coin::from_balance(house_balance, ctx);
+            transfer::transfer(house_coins, game.house);
+        } else {
+            assert!(!vector::is_empty(&secret), ESecretIsEmpty);
+            let hash = sha3_256(secret);
+            assert!(hash == digest::sha3_256_digest_to_bytes(&game.hashed_secret), EHashAndSecretDontMatch);
+            // extract balances and guess
+            let HouseData {house_balance, min_bet: _, max_bet: _} = option::extract(&mut game.house_data);
+            let BetData {stake, guess} = option::extract(&mut game.bet_data);
+            let first_byte = vector::borrow(&secret, 0);
+            let won = guess == *first_byte % 2;
+
+            if (won) {
+                let outcome = Outcome {
+                    secret,
+                    guess,
+                    player_won: true
+                };
+                option::fill(&mut game.outcome, outcome);
+                let player = option::extract(&mut game.player);
+                pay_player(player, game.house, stake, house_balance, ctx);
+            } else {
+                // player lost
+                let outcome = Outcome {
+                    secret,
+                    guess,
+                    player_won: false
+                };
+                option::fill(&mut game.outcome, outcome);
+
+                balance::join(&mut house_balance, stake);
+                transfer::transfer(coin::from_balance(house_balance, ctx), game.house);
+            };
+        }
+    }
+
+    /// Called by anyone after the required epochs have passed and house has not revealed the secret.
+    /// The house automatically loses.
+    /// Check EpochsCancelAfter for the number of epochs required to pass after `fun bet` has been called
```

Comment:
generally prefer adding full stops to any comment (this applies to the whole PR)